### PR TITLE
Handle missing context gracefully

### PR DESCRIPTION
**Pull request by jkillian**
Makes the `<Flag />` component handle a missing `<FlagsProvider />` ancestor gracefully. The component will log a warning and default to acting as if the given flag is disabled.

I believe this change should align well with the libraries current behaviors about invalid flag names. Let me know any thoughts!

### CODE REVIEW

**Comment by garbles** [comment]
Hey @JKillian, just so I understand a little better, what problem is this solving for you?

**Comment by jkillian** [comment]
It's a bit of and odd situation, but we have multiple apps that use some shared components. Some of these apps support the idea of feature flags, and some don't. We want our shared components to be able to use `<Flag />` without having to worry about if a `FlagsProvider` is provided by the app or not. In the apps that don't provide a `<FlagsProvider />`, the shared components would just default to false.

As a more theoretical argument, one of the cool things about the new React Context API is [that consumers can specify a default value even if there is no provider](https://reactjs.org/docs/context.html#contextconsumer). Why is this a good thing? I think it's a good thing because it's very difficult for a component to "reason" about what its parents are, and it's nice to have more resilient components which can function without context.

**Comment by jkillian** [comment]
FWIW, I agree this probably isn't the most common situation! But at the very least, we get a better-defined warning instead of a seemingly-arbitrary runtime error now 😄 Thanks for the quick response btw!

**Comment by garbles** [comment]
Ah ok great. This context is very helpful.

> We want our shared components to be able to use <Flag /> without having to worry about if a FlagsProvider is provided by the app or not.

So are you OK with shared components that log errors on every render? 😛

Your points about arbitrary errors and using the default value are well taken! At present the default is  of course `null`, so I think an alternative way to accommodate your use case would be the following change:

1. Allow `createFlags` to accept an optional argument which is the type of the resolved flags `T`.
2. If the optional argument is passed, then components would just use it and we're done.
3. If no optional argument is passed, then `null` (or a Symbol) is used to indicate "no default values".
4. If fetching the context in `useFlags` returns `null` (or the Symbol) then it should _throw_ a message instead of warning (because this situation is unrecoverable).
5. `useFlag` would also throw because it relies on `useFlags`. I don't think the user should have to type-check away `undefined` on a value that _should_ be present. _Maybe_ `useFlag` could have an optional second input which is the default value and is used when the context is `null` (or the Symbol). This would ensure that a value of the expected type is returned without the need to wrap it in context.

Does that make sense?

**Comment by jkillian** [comment]
> So are you OK with shared components that log errors on every render? 😛

Haha good point, if we kept this, I'd change it to use `useEffect` and only log it "on mount". But even with that change, point taken.

But, I think what you suggest seems like a good change. I'm totally fine with the following (which I think is exactly what you said):

* If `createFlags` is passed a default, when a provider is not present, `Flag`, `useFlags`, and `useFlag` will all use that passed in default and not throw.
* If `createFlags` is _not_ passed a default, and a provider is not present, `Flag`, `useFlags`, and `useFlag` will all throw (likely via calling `useFlags`).

In other words, the default behavior today doesn't really change, except that a nicer error is thrown. But, we add on a new optional behavior to let users specify a default value and prevent error-throwing.

Seems 💯 to me from a back-compat perspective and a functionality perspective!

**Comment by jkillian** [comment]
@garbles would you like me to take a stab at implementing this? Or were you planning to? I know it's less overhead overall if you want to do it yourself (and easier for me lol), but happy to redo this PR if you desire it!

**Comment by garbles** [comment]
I can take care of it in the next few days! 🤗

**Comment by garbles** [comment]
Solved by 82ef0253bbb77cf4bd0b7e23923ccfc129b650de

Made a pre-release (`yarn add flag@4.4.0-0`), that includes it but still need to update the README.

**Comment by jkillian** [comment]
Amazing - just tried out the new release, seems to be working well for us! Thanks for your work @garbles! Going to close out this PR since it's not really relevant anymore
